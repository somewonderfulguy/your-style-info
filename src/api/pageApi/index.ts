export * from './pageTypes'
export * from './pageApi'
