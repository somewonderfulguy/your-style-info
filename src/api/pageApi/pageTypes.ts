export type PageType = {
  id: string
  header: string
  components: AnyComponent[]
}

export enum ComponentType {
  // keep in alphabetical order
  Heading = 'heading',
  Image = 'image',
  Text = 'text',
  TilesCarousel = 'tilesCarousel'
}

export type AnyComponent =
  | HeadingComponent
  | ImageComponent
  | TextComponent
  | TilesCarouselComponent

export type BaseComponent = {
  type: ComponentType
  contentWidth?: 'content' | 'popout' | 'feature' | 'full'
}

export type ComponentPropsMapping = {
  // keep in alphabetical order
  [ComponentType.Heading]: Omit<HeadingComponent, 'type'>
  [ComponentType.Image]: Omit<ImageComponent, 'type'>
  [ComponentType.Text]: Omit<TextComponent, 'type'>
  [ComponentType.TilesCarousel]: Omit<TilesCarouselComponent, 'type'>
}

// keep all components in alphabetical order

export type HeadingComponent = BaseComponent & {
  type: ComponentType.Heading
  children: string
  level: 1 | 2 | 3 | 4 | 5 | 6
}

export type ImageComponent = BaseComponent & {
  type: ComponentType.Image
  url: string
  lowresBase64: string | null
  alt: string
  caption: string
  height: number
  width: number | string
}

export type TextComponent = BaseComponent & {
  type: ComponentType.Text
  children: string
}

export type TilesCarouselComponent = BaseComponent & {
  type: ComponentType.TilesCarousel
  items: []
}
