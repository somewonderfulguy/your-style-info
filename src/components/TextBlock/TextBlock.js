import React from 'react'
import {string} from 'prop-types'

const propTypes = {
  text: string
}

const defaultProps = {
  text: ''
}

const TextBlock = ({text}) => (
  <p>{text}</p>
)

TextBlock.propTypes = propTypes
TextBlock.defaultProps = defaultProps

export default TextBlock