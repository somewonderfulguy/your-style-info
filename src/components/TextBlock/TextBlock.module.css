.textBlock {
  line-height: 32px;
  font-size: 20px;
  letter-spacing: -0.003em;
  margin-top: 0;
  margin-bottom: 2em;
}

.textBlockMobile {
  composes: textBlock;
  line-height: 28px;
  font-size: 16px;
  margin-bottom: 1.5em;
}
