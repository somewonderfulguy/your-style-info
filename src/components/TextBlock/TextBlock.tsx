import { useIsDesktop } from '~contexts/screenDimensionsContext'

import styles from './TextBlock.module.css'

type Props = { text?: string }

const TextBlock = ({ text = '' }: Props) => {
  const isDesktop = useIsDesktop()
  return (
    <p className={isDesktop ? styles.textBlock : styles.textBlockMobile}>
      {text}
    </p>
  )
}

export default TextBlock
