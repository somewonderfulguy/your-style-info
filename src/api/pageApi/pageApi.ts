import { useQuery, UseQueryOptions } from '@tanstack/react-query'

import { request } from '../request'
import { pagePathPair } from '../apiConstants'
import { PageType } from './pageTypes'

const getPage = (url: string) => request<PageType>(url)

export const usePageQuery = (
  rawPath: string,
  options: Omit<UseQueryOptions<PageType>, 'queryKey'> = {}
) => {
  const path =
    rawPath[rawPath.length - 1] === '/' ? rawPath.slice(0, -1) : rawPath
  const url = pagePathPair.get(path)
  // if(!url) throw new Error(`unexpected path: ${path}`) // TODO display such errors in UI so the user know that his request didn't work

  const TEMP_URL = url?.substring(1) // will not be needed when it comes to real api requests (not local)
  return useQuery({
    queryFn: () => getPage(TEMP_URL ?? ''),
    queryKey: ['page', path],
    enabled: !!url,
    ...options
  })
}
