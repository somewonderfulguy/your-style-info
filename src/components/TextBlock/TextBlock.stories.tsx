import React from 'react'

import { ComponentStory, ComponentMeta } from '@storybook/react'

import TextBlock from './TextBlock'

const componentMeta: ComponentMeta<typeof TextBlock> = {
  title: 'TextBlock',
  component: TextBlock
}
export default componentMeta

const Template: ComponentStory<typeof TextBlock> = (args) => <TextBlock {...args} />

export const TextBlockDefault = Template.bind({})

TextBlockDefault.args = {
  text: 'Evil is Evil. Lesser, greater, middling… Makes no difference. The degree is arbitrary. The definition\'s blurred. If I\'m to choose between one evil and another… I\'d rather not choose at all.'
}