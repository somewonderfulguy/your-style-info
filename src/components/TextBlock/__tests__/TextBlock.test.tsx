import React from 'react'
import { render } from '@testing-library/react'
import { TextComponent } from '~api/pageApi'

import TextBlock from '..'

test('should match snapshot', () => {
  const { asFragment } = render(
    <TextBlock type={'text' as TextComponent['type']}>
      {`Buying cheap shoes is not money saving. It's madness.`}
    </TextBlock>
  )
  expect(asFragment()).toMatchSnapshot()
})
