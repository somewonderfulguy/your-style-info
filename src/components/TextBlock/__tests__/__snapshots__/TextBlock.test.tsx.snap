// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <p>
    Buying cheap shoes is not money saving. It's madness.
  </p>
</DocumentFragment>
`;
